### PR TITLE
[SPARK-37423][PYTHON] Inline type hints for fpm.py in python/pyspark/mllib

### PR DESCRIPTION
**Pull request by dchvn**
### What changes were proposed in this pull request?
Inline type hints for fpm.py, test.py in python/pyspark/mllib/

### Why are the changes needed?
We can take advantage of static type checking within the functions by inlining the type hints.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Existing tests

### CODE REVIEW

**Comment by dchvn** [comment]
CC @zero323 @HyukjinKwon @ueshin, please take a look if you have time. Thanks!

**Comment by dchvn** [comment]
Ping @zero323 , Very glad if you help me resolve this PR! TIA :smiley:

**Review by zero323 — COMMENTED** [review]
First of all, let's resync this with main branch.

Also,  we should properly annotate result classes. We won't be able to use `Generics`, but it is still an improvement:

```python
diff --git a/python/pyspark/mllib/fpm.py b/python/pyspark/mllib/fpm.py
index 6cff32d6ee..afd60659a0 100644
--- a/python/pyspark/mllib/fpm.py
+++ b/python/pyspark/mllib/fpm.py
@@ -17,8 +17,7 @@
 
 import sys
 
-from collections import namedtuple
-from typing import Generic, List, TypeVar
+from typing import Any, Generic, List, NamedTuple, TypeVar
 
 from pyspark import since, SparkContext
 from pyspark.mllib.common import JavaModelWrapper, callMLlibFunc
@@ -53,7 +52,7 @@ class FPGrowthModel(JavaModelWrapper, JavaSaveable, JavaLoader["FPGrowthModel"],
     """
 
     @since("1.4.0")
-    def freqItemsets(self) -> RDD["FPGrowth.FreqItemset[T]"]:  # type: ignore[misc]
+    def freqItemsets(self) -> RDD["FPGrowth.FreqItemset"]:
         """
         Returns the frequent itemsets of this model.
         """
@@ -65,7 +64,7 @@ class FPGrowthModel(JavaModelWrapper, JavaSaveable, JavaLoader["FPGrowthModel"],
         """
         Load a model from the given path.
         """
-        model = cls._load_java(sc, path)  # type: ignore[attr-defined]
+        model = cls._load_java(sc, path)
         assert sc._jvm is not None
         wrapper = sc._jvm.org.apache.spark.mllib.api.python.FPGrowthModelWrapper(model)
         return FPGrowthModel(wrapper)
@@ -102,13 +101,16 @@ class FPGrowth:
         model = callMLlibFunc("trainFPGrowthModel", data, float(minSupport), int(numPartitions))
         return FPGrowthModel(model)
 
-    class FreqItemset(Generic[T], namedtuple("FreqItemset", ["items", "freq"])):
+    class FreqItemset(NamedTuple):
         """
         Represents an (items, freq) tuple.
 
         .. versionadded:: 1.4.0
         """
 
+        items: List[Any]
+        freq: int
+
 
 @inherit_doc
 class PrefixSpanModel(JavaModelWrapper, Generic[T]):
@@ -131,7 +133,7 @@ class PrefixSpanModel(JavaModelWrapper, Generic[T]):
     """
 
     @since("1.6.0")
-    def freqSequences(self) -> RDD["PrefixSpan.FreqSequence[T]"]:  # type: ignore[misc]
+    def freqSequences(self) -> RDD["PrefixSpan.FreqSequence"]:
         """Gets frequent sequences"""
         return self.call("getFreqSequences").map(lambda x: PrefixSpan.FreqSequence(x[0], x[1]))
 
@@ -190,13 +192,16 @@ class PrefixSpan:
         )
         return PrefixSpanModel(model)
 
-    class FreqSequence(Generic[T], namedtuple("FreqSequence", ["sequence", "freq"])):
+    class FreqSequence(NamedTuple):
         """
         Represents a (sequence, freq) tuple.
 
         .. versionadded:: 1.6.0
         """
 
+        sequence: List[List[Any]]
+        freq: int
+
 
 def _test() -> None:
     import doctest
```

File: python/pyspark/mllib/fpm.py [hunk 1]
```diff
@@ -17,17 +17,20 @@
...
+from typing import Any, Generic, List, NamedTuple, TypeVar
 
-from pyspark import since
+from pyspark import since, SparkContext
 from pyspark.mllib.common import JavaModelWrapper, callMLlibFunc
 from pyspark.mllib.util import JavaSaveable, JavaLoader, inherit_doc
+from pyspark.rdd import RDD
 
-__all__ = ["FPGrowth", "FPGrowthModel", "PrefixSpan", "PrefixSpanModel"]
+_all__ = ["FPGrowth", "FPGrowthModel", "PrefixSpan", "PrefixSpanModel"]
```
Comment (itholic) [inline python/pyspark/mllib/fpm.py:27]: mistake?

Comment (dchvn) [reply python/pyspark/mllib/fpm.py]: Oh, my mistake, updated! Thanks

**Review by zero323 — APPROVED** [review]
Looks OK. 

cc @HyukjinKwon  @ueshin  ‒ do you think it is acceptable to backport this and the remaining ones to branch-3.3?

**Review by itholic — APPROVED** [review]
Looks good

**Comment by zero323** [comment]
Merged into master and branch-3.3.

**Pull request closed without merging**